IdentityCore/src/validation/MSIDCIAMAuthority.m

```diff
@@ -0,0 +1,184 @@
+#import "MSIDJsonSerializableFactory.h"
+#import "MSIDJsonSerializableTypes.h"
+#import "MSIDProviderType.h"
+#import "MSIDAADTenant.h"
+#import "MSIDAADAuthority.h"
+#import "NSString+MSIDExtensions.h"
+
+@implementation MSIDCIAMAuthority
+
++ (void)load
+{
+    [MSIDJsonSerializableFactory registerClass:self forClassType:MSID_JSON_TYPE_CIAM_AUTHORITY];
+    [MSIDJsonSerializableFactory mapJSONKey:MSID_PROVIDER_TYPE_JSON_KEY keyValue:MSID_JSON_TYPE_PROVIDER_CIAM kindOfClass:MSIDAuthority.class toClassType:MSID_JSON_TYPE_CIAM_AUTHORITY];
+}
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                           rawTenant:(nullable NSString *)rawTenant
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{    
+    self = [self initWithURL:url validateFormat:validateFormat context:context error:error];
+    if (self)
+    {
+        if (rawTenant)
+        {
+            if ([self.class isAuthorityFormatValid:url context:context error:nil])
+            {
+                _url = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [url msidHostWithPortIfNecessary], rawTenant]];
+                _realm = rawTenant;
+            }
+        }
+    }
+
+    return self;
+}
+
+- (nullable instancetype)initWithURL:(NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                             context:(id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{
+    self = [super initWithURL:url validateFormat:validateFormat context:context error:error];
+    
+    NSArray *hostComponents = [url.msidHostWithPortIfNecessary componentsSeparatedByString:@"."];
+    
+    //If we have the URL tenant.ciamlogin.com
+    if (url.pathComponents.count == 0 || [[url lastPathComponent] isEqual:@"/"])
+    {
+        url = [url URLByAppendingPathComponent:hostComponents[0]];
+        url = [NSURL URLWithString:[url.absoluteString stringByAppendingString:@".onmicrosoft.com"]];
+    }
+    
+    if (self)
+    {
+        _url = url;
+        _url = [MSIDAADAuthority normalizedAuthorityUrl:url context:context error:error];
+        if (!_url) return nil;
+        self.url = url;
+    }
+    
+    return self;
+}
+    
+- (instancetype)initWithURL:(NSURL *)url
+                    context:(id<MSIDRequestContext>)context
+                      error:(NSError **)error
+{
+    return [self initWithURL:url validateFormat:YES context:context error:error];
+}
+
++ (BOOL)isAuthorityFormatValid:(NSURL *)url
+                       context:(id<MSIDRequestContext>)context
+                         error:(NSError **)error
+{
+    if (![super isAuthorityFormatValid:url context:context error:error]) return NO;
+    
+    NSArray *hostComponents = [url.msidHostWithPortIfNecessary componentsSeparatedByString:@"."];
+    NSString *ciamTenant = hostComponents[1];
```

Comment:
This will crash if there're no hostcomponents, please move the check from line 111 above this one